pyo3-macros-backend/src/introspection.rs

```diff
@@ -0,0 +1,227 @@
+use syn::{Attribute, Ident};
+
+static GLOBAL_COUNTER_FOR_UNIQUE_NAMES: AtomicUsize = AtomicUsize::new(0);
+
+pub fn module_introspection_code<'a>(
+    pyo3_crate_path: &PyO3CratePath,
+    name: &str,
+    members: impl IntoIterator<Item = &'a Ident>,
+    members_cfg_attrs: impl IntoIterator<Item = &'a Vec<Attribute>>,
+) -> TokenStream {
+    let stub = IntrospectionNode::Map(
+        [
+            ("type", IntrospectionNode::String("module")),
+            ("id", IntrospectionNode::IntrospectionId(None)),
+            ("name", IntrospectionNode::String(name)),
+            (
+                "members",
+                IntrospectionNode::List(
+                    members
+                        .into_iter()
+                        .zip(members_cfg_attrs)
+                        .filter_map(|(member, attributes)| {
+                            if attributes.is_empty() {
+                                Some(IntrospectionNode::IntrospectionId(Some(member)))
+                            } else {
+                                None // TODO: properly interpret cfg attributes
+                            }
+                        })
+                        .collect(),
+                ),
+            ),
+        ]
+        .into(),
+    )
+    .emit(pyo3_crate_path);
+    let introspection_id = introspection_id_const();
+    quote! {
+        #stub
+        #introspection_id
+    }
+}
+
+pub fn class_introspection_code(
+    pyo3_crate_path: &PyO3CratePath,
+    ident: &Ident,
+    name: &str,
+) -> TokenStream {
+    let stub = IntrospectionNode::Map(
+        [
+            ("type", IntrospectionNode::String("class")),
+            ("id", IntrospectionNode::IntrospectionId(Some(ident))),
+            ("name", IntrospectionNode::String(name)),
+        ]
+        .into(),
+    )
+    .emit(pyo3_crate_path);
+    let introspection_id = introspection_id_const();
+    quote! {
+        #stub
+        impl #ident {
+            #introspection_id
+        }
+    }
+}
+
+pub fn function_introspection_code(pyo3_crate_path: &PyO3CratePath, name: &str) -> TokenStream {
+    let stub = IntrospectionNode::Map(
+        [
+            ("type", IntrospectionNode::String("function")),
+            ("id", IntrospectionNode::IntrospectionId(None)),
+            ("name", IntrospectionNode::String(name)),
+        ]
+        .into(),
+    )
+    .emit(pyo3_crate_path);
+    let introspection_id = introspection_id_const();
+    quote! {
+        #stub
+        #introspection_id
+    }
+}
+
+enum IntrospectionNode<'a> {
+    String(&'a str),
+    IntrospectionId(Option<&'a Ident>),
+    Map(HashMap<&'static str, IntrospectionNode<'a>>),
+    List(Vec<IntrospectionNode<'a>>),
+}
+
+impl IntrospectionNode<'_> {
+    fn emit(&self, pyo3_crate_path: &PyO3CratePath) -> TokenStream {
+        let mut content = ConcatenationBuilder::default();
+        self.add_to_serialization(&mut content);
+        let content = content.into_token_stream(pyo3_crate_path);
```

Comment:
I seems to me that the content is defined to be JSON payloads matching the `Chunk` enum in pyo3-introspection? Maybe shall we document / link to that, it took me a while to find...

Reply:
Indeed. I have added a line to the module doc comments. is it good enough?